cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/SourceControlManagementService.java

```diff
@@ -169,4 +187,69 @@ public PushAppResponse pushApp(ApplicationReference appRef, @Nullable String com
 
     return pushResponse;
   }
+
+  /**
+   * Pull the application from linked repository and deploy it in current namespace.
+   * @param appRef application reference to deploy with
+   * @return {@link ApplicationRecord} of the deployed application.
+   * @throws Exception when {@link ApplicationLifecycleService} fails to deploy.
+   */
+  public ApplicationRecord pullAndDeploy(ApplicationReference appRef) throws Exception {
+    PullAppResponse<?> pullResponse = pullAndValidateApplication(appRef);
+
+    AppRequest<?> appRequest = pullResponse.getAppRequest();
+    KerberosPrincipalId ownerPrincipalId =
+      appRequest.getOwnerPrincipal() == null ? null : new KerberosPrincipalId(appRequest.getOwnerPrincipal());
+
+    // if we don't null check, it gets serialized to "null"
+    Object config = appRequest.getConfig();
+    String configString = config == null ? null : config instanceof String ? (String) config : GSON.toJson(config);
```

Comment:
shouldn't need this instanceOf check.

Also, I would imagine this is duplicate code, are we doing basically the same thing in the AppLifecycleHttpHandler? If so it may be good to consolidate and push it into the deployApp() method.

Reply:
yes, this is basically the same from `AppLifecycleHttpHandler`, will push them into `deployApp`

Reply:
@albertshau Interestingly this `instanceOf check` is needed. This is a bit tricky that if the config is already a String and we still do `GSON.toJson(config)`, it adds extra quotes and downstream deploy will fail because of invalid json error